theseus/src/api/modpacks.rs

```diff
@@ -0,0 +1,231 @@
+#[serde(rename_all = "camelCase")]
+struct PackFormat {
+    pub game: String,
+    pub format_version: i32,
+    pub version_id: String,
+    pub name: String,
+    pub summary: Option<String>,
+    pub files: Vec<PackFile>,
+    pub dependencies: HashMap<PackDependency, String>,
+}
+
+#[derive(Serialize, Deserialize, Eq, PartialEq)]
+#[serde(rename_all = "camelCase")]
+struct PackFile {
+    pub path: String,
+    pub hashes: HashMap<PackFileHash, String>,
+    pub env: Option<HashMap<EnvType, SideType>>,
+    pub downloads: Vec<String>,
+    pub file_size: u32,
+}
+
+#[derive(Serialize, Deserialize, Eq, PartialEq, Hash)]
+#[serde(rename_all = "camelCase", from = "String")]
+enum PackFileHash {
+    Sha1,
+    Sha512,
+    Unknown(String),
+}
+
+impl From<String> for PackFileHash {
+    fn from(s: String) -> Self {
+        return match s.as_str() {
+            "sha1" => PackFileHash::Sha1,
+            "sha512" => PackFileHash::Sha512,
+            _ => PackFileHash::Unknown(s),
+        };
+    }
+}
+
+#[derive(Serialize, Deserialize, Eq, PartialEq, Hash)]
+#[serde(rename_all = "camelCase")]
+enum EnvType {
+    Client,
+    Server,
+}
+
+#[derive(Serialize, Deserialize, Clone, Hash, PartialEq, Eq)]
+#[serde(rename_all = "kebab-case")]
+enum PackDependency {
+    Forge,
+    FabricLoader,
+    QuiltLoader,
+    Minecraft,
+}
+
+pub async fn install_modpack_from_version_id(
+    version_id: String,
+) -> crate::Result<PathBuf> {
+    let version: ModrinthVersion = REQWEST_CLIENT
+        .get(format!("{}version/{}", MODRINTH_API_URL, version_id))
+        .send()
+        .await?
+        .json()
+        .await?;
+
+    let (url, hash) =
+        if let Some(file) = version.files.iter().find(|x| x.primary) {
+            Some((file.url.clone(), file.hashes.get("sha1")))
+        } else if let Some(file) = version.files.first() {
+            Some((file.url.clone(), file.hashes.get("sha1")))
+        } else {
+            None
+        }
+        .ok_or_else(|| {
+            crate::ErrorKind::InputError(
+                "Specified version has no files".to_string(),
+            )
+        })?;
+
+    let file = async {
+        let state = &State::get().await?;
+        let semaphore = state.io_semaphore.acquire().await?;
+        fetch(&*url, hash.map(|x| &**x), &semaphore).await
+    }
+    .await?;
+
+    // TODO: fetch icon and pass it here
+    install_pack(file, None).await
+}
+//
+// pub fn install_modpack_from_file(path: PathBuf) -> crate::Result<()> {}
+
+async fn install_pack(
+    file: bytes::Bytes,
+    icon: Option<PathBuf>,
+) -> crate::Result<PathBuf> {
+    let state = &State::get().await?;
+
+    let reader = Cursor::new(file);
+    let mut zip = ZipArchive::new(reader).map_err(|_| {
+        crate::Error::from(crate::ErrorKind::InputError(
+            "Failed to read input modpack zip".to_string(),
+        ))
+    })?;
+
+    let index_json = zip.by_name("modrinth.index.json");
```

Comment:
I imagine this filename will be used elsewhere for modpack stuff as we add more functionality- could be worth making a constant

Reply:
hmm I don't imagine it actually since the manifest is only used during the install/updating- which will be similar functions